Mathlib/Tactic/Algebraize.lean

```diff
@@ -0,0 +1,263 @@
+
+/-!
+
+## Algebraize tactic
+
+This file defines the `algebraize` tactic. The basic functionality of this tactic is to
+automatically add `Algebra` instances given `RingHom`s. For example, `algebraize f g` where
+`f : A →+* B` and `g : B →+* C` are `RinHom`s, will add the instances `Algebra A B` and
+`Algebra B C` corresponding to these `RingHom`s.
+
+## Further functionality
+
+When given a composition of `RingHom`s, e.g. `algebraize g.comp f`, the tactic will also try to
+add the instance `IsScalarTower A B C` if possible.
+
+After having added suitable `Algebra` and `IsScalarTower` instances, the tactic will search through
+the local context for `RingHom` properties that can be converted to properties of the corresponding
+`Algebra`. For example, given `f : A →+* B` and `hf : f.FiniteType`, then `algebraize f` will add
+the instance `Algebra A B` and the corresponding property `Algebra.FiniteType A B`. The tactic knows
+which `RingHom` properties have a corresponding `Algebra` property through the `algebraize`
+attribute.
+
+## Algebraize attribute
+
+The `algebraize` attribute is used to tag `RingHom` properties that can be converted to `Algebra`
+properties. It assumes that the tagged declaration has a name of the form `RingHom.Property` and
+that the corresponding `Algebra` property has the name `Algebra.Property`.
+
+If not, the `Name` of the corresponding algebra property can be provided as optional argument. The
+specified declaration should be one of the following:
+
+1. An inductive type (i.e. the `Algebra` property itself), in this case it is assumed that the
+`RingHom` and the `Algebra` property are definitionally the same, and the tactic will construct the
+`Algebra` property by giving the `RingHom` property as a term.
+
+2. A constructor for the `Algebra` property. In this case it is assumed that the `RingHom` property
+is the last argument of the constructor, and that no other explicit argument is needed. The tactic
+then constructs the `Algebra` property by applying the constructor to the `RingHom` property.
+
+Here are three examples of properties tagged with the `algebraize` attribute:
+```
+@[algebraize]
+def RingHom.FiniteType (f : A →+* B) : Prop :=
+  @Algebra.FiniteType A B _ _ f.toAlgebra
+```
+An example when the `Name` is provided (as the `Algebra` does not have the expected name):
+```
+@[algebraize Module.Finite]
+def Module.Finite (f : A →+* B) : Prop :=
+  letI : Algebra A B := f.toAlgebra
+  Module.Finite A B
+```
+An example with a constructor as parameter (as the two properties are not definitonally the same):
+```
+@[algebraize Algebra.Flat.out]
+class RingHom.Flat {R : Type u} {S : Type v} [CommRing R] [CommRing S] (f : R →+* S) : Prop where
+  out : f.toAlgebra.Flat := by infer_instance
+```
+
+## Algebraize'
+
+To avoid searching through the local context and adding corresponding `Algebra` properties, use
+`algebraize'` which only adds `Algebra` and `IsScalarTower` instances.
```

Comment:
```
Version of algebraize, which only adds Algebra instances and IsScalarTower instances.

macro_rules
  | `(tactic| algebraize' $[$t:term]*) =>
    `(tactic| algebraize (config := {searchContext := false}) $t*)
```

Since `algebraize'` is literally just an abbreviation for `algebraize (config := {searchContext := false})`  I would argue it would be fine to leave the macro in as is. If it's any more verbose, one could directly use the version with the `config` instead.

To throw in another suggestion, one could rename `searchContext` to one of `properties`, `findProperties`, or `searchProperties`. Especially `algebraize (config := {properties := false})` isn't that verbose anymore.

In any case, I think @callesonne it's reasonable if you have another think about the naming of macro+config option (maybe with your co-authors) and then decide on what you like best :)

Reply:
@joneugster I really like the config name option suggestion!

How about `algebraize_only` instead of `algebraize'`, or somehow making the syntax `algebraize only` work?

Reply:
My suggestion is to use the suggestion `algebraize_only`. That sounds reasonable.

With `algebraize only` I would argue that `algebraize only [f, g, f.comp g]` somehow looks unexpected. For me the `only`-keyword would suggest that whatever follows is a subset of all tagged declarations, i.e. I'd rather expect `algebraize only [RingHom.Finite]` to work, which might not integrate well: `algebraize only [RingHom.Finite] at f g (f.comp g)` would be uniform with other tactics, but then we're back at the sequence of terms... (on a techincal level, it wouldn't be difficult: one could directly [copy the syntax from `simp`](https://github.com/leanprover/lean4/blob/e9e858a4484905a0bfe97c4f05c3924ead02eed8/src/Init/Tactics.lean#L570-L571))

Reply:
That would be very cool to implement! @kim-em what do you think about this `algebraize only [...] at ...` syntax as in Jon's new suggestion? Or maybe I just call the macro `algebraize_only` for now, so that some version of this reaches mathlib, and then work on `algebraize only` after.

Reply:
I think the problem is that usually the `at ...` is essentially a sequence of `ident`s, but here it would be an undeterminated sequence of `term`s, which is much more general and the reason Kim was worried about parsing errors (if I understand correctly). So this would require thinking again about the way `f.comp g` is handled. This might be out-of-scope for this PR.

For the sake of preserving your own efforts, I would say just use `algebraize_only` here and work this out in a follow-up PR if you have the motivation. We already kept adding more and more, and I think it would be better to get a first version merged and then have improvements in follow-up PRs.

Reply:
switched to `algebraize_only` and config options`properties`. for now